File: Sorting/ColoringSocks.cpp
/*	Turner Atwood
 *	3/25/19
 *	Coloring Socks [3.1] : (https://open.kattis.com/problems/color)
 *	Count the socks of each color and create sliding windows to add them to washers
 */

#include <iostream>
#include <map>

using namespace std;

int main() {
	// Input
	map<int,int> numbers;
	int S,C,K;
	cin >> S >> C >> K;
	int num;
	// Add the count of each sock to the map
	for (int i = 0; i < S; i++) {
		cin >> num;
		if (numbers.count(num)) {
			numbers[num]++;
		}
		else {
			numbers[num] = 1;
		}
	}

	// Init variables
	int washers = 0;
	int leftovers = 0;
	int thisColor,count;
	int lastColor = 0;

	// Try socks for every color (least to greatest)
	for (map<int,int>::iterator it = numbers.begin(); it != numbers.end(); it++) {
		thisColor = it->first;
		count = it->second;
		// If the last washer has room
		if (leftovers != 0) {
			// Check to see if these socks can go in it
			if (thisColor - lastColor <= K) {
				count += leftovers;
			}
			// If not, start a new washer
			else {
				washers++;
			}
		}
		// Fill all possible washers 
		washers  += count / C;
		leftovers = count % C;
		lastColor = thisColor;
	}
	// Include any remaining socks in a final washer
	if (leftovers > 0) {
		washers++;
	}

	// Output
	cout << washers << endl;

}

File: Sorting/ColoringSocks.py
"""
/*	Turner Atwood
 *	3/25/19
 *	Coloring Socks [2.5] https://open.kattis.com/problems/color
 *	Sort the socks then Ad-hoc
 */
"""

def main():
	S,C,K = [int(i) for i in input().split()]
	socks = [int(i) for i in input().split()]
	socks.sort()

	machines = 0
	cur_color = socks[0]
	cur_cap  = 0
	for i in range(S):
		# Check to see if the machine is full
		if cur_cap == C:
			machines += 1
			cur_cap = 0

		sock = socks[i]

		# If the machine is empty, change the color to the current sock
		if cur_cap == 0:
			cur_color = sock

		# If the new sock is close enough to the original, throw it in
		if abs(sock-cur_color) <= K:
			cur_cap += 1
		# If not, start a new machine with that color
		else:
			machines += 1
			cur_cap = 1
			cur_color = sock

	# Count another machine if the last one is unfinished
	if cur_cap > 0:
		machines += 1
	print(machines)


if __name__ == "__main__":
	main()